[JSC] The fast path for JSONP should handle TDZ

hyjorc1 opened the pull request:
#### 0f6d9985ade696e37ec5653318935961e3f73056
<pre>
[JSC] The fast path for JSONP should handle TDZ
<a href="https://bugs.webkit.org/show_bug.cgi?id=243578">https://bugs.webkit.org/show_bug.cgi?id=243578</a>
rdar://98106247

Reviewed by NOBODY (OOPS!).

All the fast paths for JSONP should throw exceptions for handling TDZ.

* JSTests/stress/jsonp-program-evaluate-path-must-consider-global-lexical-environment.js:
* JSTests/stress/jsonp-program-evaluate-path-must-handle-tdz.js: Added.
(shouldThrow):
(test1):
(test2):
(test3):
(test4):
(test5):
(test6):
(test7):
(test8):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeProgram):
</pre>

MenloDorian commented on Source/JavaScriptCore/jit/JITOperations.cpp:
```diff
@@ -1760,12 +1760,14 @@ JSC_DEFINE_JIT_OPERATION(operationNewAsyncGeneratorFunctionWithInvalidatedReallo
 JSC_DEFINE_JIT_OPERATION(operationSetFunctionName, void, (JSGlobalObject* globalObject, JSCell* funcCell, EncodedJSValue encodedName))
 {
     VM& vm = globalObject->vm();
+    auto throwScope = DECLARE_THROW_SCOPE(vm);
```
Don't need this because you don't have any exception check (see below) or and you don't throw any exceptions.

justinmichaud commented on Source/JavaScriptCore/jit/JITOperations.cpp:
```diff
@@ -1760,12 +1760,14 @@ JSC_DEFINE_JIT_OPERATION(operationNewAsyncGeneratorFunctionWithInvalidatedReallo
     JSFunction* func = jsCast<JSFunction*>(funcCell);
     JSValue name = JSValue::decode(encodedName);
     func->setFunctionName(globalObject, name);
+    RETURN_IF_EXCEPTION(throwScope, void());
```
Release_and_return?

justinmichaud replied:
Either this scope is not needed since we only call one thing that can throw, or we are missing an exception check here

MenloDorian commented on Source/JavaScriptCore/jit/JITOperations.cpp:
```diff
@@ -1760,12 +1760,14 @@ JSC_DEFINE_JIT_OPERATION(operationNewAsyncGeneratorFunctionWithInvalidatedReallo
     JSFunction* func = jsCast<JSFunction*>(funcCell);
     JSValue name = JSValue::decode(encodedName);
     func->setFunctionName(globalObject, name);
+    RETURN_IF_EXCEPTION(throwScope, void());
```
Don't need this since you're just going to return anyway.  It's the caller's responsibility to handle the exception now.

justinmichaud commented on Source/JavaScriptCore/runtime/JSFunction.cpp:
```diff
@@ -576,11 +577,12 @@ void JSFunction::reifyName(VM& vm, JSGlobalObject* globalObject)
     else
         name = ecmaName.string();
-    reifyName(vm, globalObject, name);
+    RELEASE_AND_RETURN(throwScope, reifyName(vm, globalObject, name));
```
ditto

MenloDorian replied:
Don't need this.

justinmichaud commented on Source/JavaScriptCore/runtime/JSFunction.cpp:
```diff
@@ -597,9 +599,14 @@ void JSFunction::reifyName(VM& vm, JSGlobalObject* globalObject, String name)
+        name = tryMakeString("set ", name);
+
+    if (!name) {
+        throwOutOfMemoryError(globalObject, throwScope, "Invalid name length"_s);
```
Maybe we can be more descriptive?

MenloDorian replied:
I suggest either not providing a string description, or having it say something that can actually help the developer understand the issue and fix it.  "Invalid name length" doesn't tell the developer much.  Something like "Function name is too long" would actually give a clue as to where the issue is and how to fix it.

MenloDorian commented on Source/JavaScriptCore/runtime/JSFunction.cpp:
```diff
@@ -567,6 +567,7 @@ void JSFunction::reifyLength(VM& vm)
 
 void JSFunction::reifyName(VM& vm, JSGlobalObject* globalObject)
 {
+    auto throwScope = DECLARE_THROW_SCOPE(vm);
```
Don't need this.

justinmichaud reviewed (commented):
Let's see what EWS says

MenloDorian commented on Source/JavaScriptCore/runtime/JSFunction.cpp:
```diff
@@ -597,9 +598,14 @@ void JSFunction::reifyName(VM& vm, JSGlobalObject* globalObject, String name)
+        name = tryMakeString("set ", name);
+
+    if (!name) {
+        throwOutOfMemoryError(globalObject, throwScope, "makeString overflowed"_s);
```
"makeString overflowed" is definitely the wrong error message to use.  The message is intended for the JS developer.  `makeString` is a VM internal implementation detail.  How would they know what this is, or how to fix it, or why would they care here?  Tell the JS developer something that can help them fix the issue.  I think your "Invalid name length" message was actually pretty good.  It would be better if it tells the developer that the "name" in question came from a function, and better yet if it tells them it's from the getter or setter.  That way, they know specifically where the issue is.

MenloDorian commented on Source/JavaScriptCore/runtime/JSFunction.cpp:
```diff
@@ -597,9 +598,14 @@ void JSFunction::reifyName(VM& vm, JSGlobalObject* globalObject, String name)
-        name = makeString("set ", name);
+        name = tryMakeString("set ", name);
+
+    if (!name) {
```
nit: not sure it matters a lot, but I would add `UNLIKELY` around `!name` here because this will normally not happen.

MenloDorian commented on Source/JavaScriptCore/runtime/JSFunction.cpp:
```diff
@@ -516,6 +516,17 @@ String getCalculatedDisplayName(VM& vm, JSObject* object)
 }
 
+template<typename... StringTypes>
+String JSFunction::makeNameWithOutOfMemoryCheck(JSGlobalObject* globalObject, ThrowScope& throwScope, const char* messagePrefix, StringTypes... strings)
```
Does this need to be a member function?  It doesn't look like it relies on any private / protected members of JSFunction.

MenloDorian commented on Source/JavaScriptCore/runtime/JSFunction.h:
```diff
@@ -215,6 +215,9 @@ class JSFunction : public JSCallee {
 
+    template<typename... StringTypes>
+    String makeNameWithOutOfMemoryCheck(JSGlobalObject*, ThrowScope&, const char* messagePrefix, StringTypes... strings);
+
```
This can be removed if we make this not a member function.

MenloDorian commented on Source/JavaScriptCore/runtime/JSFunction.cpp:
```diff
@@ -661,9 +673,11 @@ JSFunction::PropertyStatus JSFunction::reifyLazyLengthIfNeeded(VM& vm, JSGlobalO
         if (!hasReifiedName()) {
             reifyName(vm, globalObject);
+            RETURN_IF_EXCEPTION(scope, PropertyStatus::Lazy);
             return PropertyStatus::Reified;
```
Currently, `reifyName` returns `void`.  Since `reifyName` is a private method, you could just make it return `PropertyStatus` (either `PropertyStatus::Lazy` or `PropertyStatus::Reified`.  With that, you can just use `RELEASE_AND_RETURN` here and obviate the need for this exception check.

MenloDorian commented on Source/JavaScriptCore/runtime/JSFunction.cpp:
```diff
@@ -682,9 +696,10 @@ JSFunction::PropertyStatus JSFunction::reifyLazyBoundNameIfNeeded(VM& vm, JSGlob
+    if (isBuiltinFunction()) {
         reifyName(vm, globalObject);
-    else if (this->inherits<JSBoundFunction>()) {
+        RETURN_IF_EXCEPTION(scope, PropertyStatus::Lazy);
```
Ditto.

MenloDorian commented on Source/JavaScriptCore/runtime/JSFunction.cpp:
```diff
@@ -535,13 +546,13 @@ void JSFunction::setFunctionName(JSGlobalObject* globalObject, JSValue value)
         RETURN_IF_EXCEPTION(scope, void());
     }
-    reifyName(vm, globalObject, name);
+    RELEASE_AND_RETURN(scope, reifyName(vm, globalObject, name));
```
If you make `reifyName` return `PropertyStatus`, you'll have to add a `(void)` cast to placate the compiler.

MenloDorian commented on Source/JavaScriptCore/runtime/JSFunction.cpp:
```diff
@@ -595,11 +607,11 @@ void JSFunction::reifyName(VM& vm, JSGlobalObject* globalObject, String name)
-        name = makeString("set ", name);
+        makeNameWithOutOfMemoryCheck(globalObject, throwScope, "Setter ", "set ", name);
 
     rareData->setHasReifiedName();
```
I just realized that this is wrong.  The patch as you have it would `setHasReifiedName` even though you fail to do so.  You should have `makeNameWithOutOfMemoryCheck` return a bool to indicate success or not.  It not successful, return with `PropertyStatus::Lazy`.  Also, to placate the exception check validator, you'll need to `ASSERT(!throwScope.exception || success)` before the `if (success)` check.

MenloDorian replied:
You can tests for this scenario by running your test cases twice e.g.
```
shouldThrow(classSetter, setterExpectedError);
shouldThrow(classSetter, setterExpectedError); // Make sure it should still throw.
```
If you didn't fix this bug, I suspect the 2nd test case invocation above till fail.

MenloDorian replied:
I think this bug is one of the reasons you're getting EWS failures.

MenloDorian commented on JSTests/stress/function-name-too-long-to-reify.js:
```diff
@@ -0,0 +1,62 @@
+shouldThrow(classGetter, getterExpectedError);
+
+shouldThrow(objectSetter, setterExpectedError);
+shouldThrow(objectGetter, getterExpectedError);
```
Please duplicate this one as well.  We would want to test the these case as well, not just the above setter one.

MenloDorian commented on Source/JavaScriptCore/runtime/JSFunction.cpp:
```diff
@@ -595,14 +609,16 @@ void JSFunction::reifyName(VM& vm, JSGlobalObject* globalObject, String name)
 
     rareData->setHasReifiedName();
     putDirect(vm, propID, jsString(vm, WTFMove(name)), initialAttributes);
+    RELEASE_AND_RETURN(throwScope, PropertyStatus::Reified);
```
You don't need `RELEASE_AND_RETURN` here.  A normal `return  PropertyStatus::Reified` works just fine.  You only want to use `RELEASE_AND_RETURN` when the thing you evaluate before returning can potentially throw an exception.

MenloDorian reviewed (approved):
r=me with changes.  Since you've added a `ThrowScope`, please also run the jsc stress tests on a Debug build to check if there are any new unhandled exceptions uncovered / introduced.  If so, please fix those as well before landing.

MenloDorian commented on JSTests/stress/function-name-too-long-to-reify.js:
```diff
@@ -0,0 +1,63 @@
+
+shouldThrow(classSetter, setterExpectedError);
+shouldThrow(classSetter, setterExpectedError); // Make sure it should still throw.
+shouldThrow(classGetter, getterExpectedError);
```
You should test this case too.

MenloDorian commented on JSTests/stress/function-name-too-long-to-reify.js:
```diff
@@ -0,0 +1,63 @@
+
+shouldThrow(objectSetter, setterExpectedError);
+shouldThrow(objectSetter, setterExpectedError); // Make sure it should still throw.
+shouldThrow(objectGetter, getterExpectedError);
```
Ditto.

hyjorc1 committed:
[JSC] JSFunction::reifyName should throw an OutOfMemoryError when reifying a too long string
https://bugs.webkit.org/show_bug.cgi?id=242538

Reviewed by Mark Lam.

JSFunction::reifyName should call WTF::tryMakeString instead of WTF::makeString
because the resultant name string may be too long and cannot be reified. In such
a case, it should throw an OutOfMemoryError.

* Source/JavaScriptCore/runtime/JSFunction.cpp:
(JSC::JSFunction::setFunctionName):
(JSC::JSFunction::reifyName):
(JSC::JSFunction::reifyLazyNameIfNeeded):
(JSC::JSFunction::reifyLazyBoundNameIfNeeded):

Canonical link: https://commits.webkit.org/252370@main
